helix-core/src/indent.rs

```diff
@@ -186,106 +188,338 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
-        Some(node) => node,
-        None => return None,
-    };
+/// The total indent for some line of code.
+/// This is usually constructed by successively adding instances of [AddedIndent]
+struct IndentResult {
+    /// The total indent (the number of indent levels).
+    /// The string that this results in depends on the indent style (spaces or tabs, etc.)
+    indent: i32,
+}
+impl IndentResult {
+    fn new() -> Self {
+        IndentResult { indent: 0 }
+    }
+    /// Add the given [AddedIndent] to the [IndentResult].
+    /// The [AddedIndent] should be the combination of all the added indents for one line.
+    fn add(&mut self, added: &AddedIndent) {
+        if added.indent && !added.outdent {
+            self.indent += 1;
+        } else if added.outdent && !added.indent {
+            self.indent = self.indent.saturating_sub(1);
+        }
+    }
+    fn as_string(&self, indent_style: &IndentStyle) -> String {
+        indent_style
+            .as_str()
+            .repeat(std::cmp::max(self.indent, 0) as usize)
+    }
+}
 
-    while let Some(parent) = node.parent() {
-        if parent.start_byte() == node.start_byte() {
-            node = parent
+// Get the node where to start the indent query (this is usually just the lowest node containing byte_pos)
+fn get_lowest_node<'a>(root: Node<'a>, _query: &IndentQuery, byte_pos: usize) -> Option<Node<'a>> {
+    root.descendant_for_byte_range(byte_pos, byte_pos)
+    // TODO Special handling for languages like python
+}
+
+// Computes for node and all ancestors whether they are the first node on their line
+// The first entry in the return value represents the root node, the last one the node itself
+fn get_first_in_line(mut node: Node, byte_pos: usize, new_line: bool) -> Vec<bool> {
+    let mut first_in_line = Vec::new();
+    loop {
+        if let Some(prev) = node.prev_sibling() {
+            // If we insert a new line, the first node at/after the cursor is considered to be the first in its line
+            let first = prev.end_position().row != node.start_position().row
+                || (new_line && node.start_byte() >= byte_pos && prev.start_byte() < byte_pos);
+            first_in_line.push(Some(first));
+        } else {
+            // Nodes that have no previous siblings are first in their line iff their parent is
+            // (which we don't know yet)
+            first_in_line.push(None);
+        }
+        if let Some(parent) = node.parent() {
+            node = parent;
         } else {
             break;
         }
     }
 
-    Some(node)
+    let mut result = Vec::with_capacity(first_in_line.len());
+    let mut parent_is_first = true; // The root node is by definition the first node in its line
+    for first in first_in_line.into_iter().rev() {
+        if let Some(first) = first {
+            result.push(first);
+            parent_is_first = first;
+        } else {
+            result.push(parent_is_first);
+        }
+    }
+    result
```

Comment:
Why not just do collect rather than manually pushing it into vec?

Reply:
How would we construct the iterator to collect from? Maybe there's some iterator function I'm missing but because we need the external state `parent_is_first`, this doesn't seem like it's nicely doable with iterators.

Reply:
```suggestion
    // The root node is by definition the first node in its line
    first_in_line
        .into_iter()
        .rev()
        .scan(true, |parent_is_first, first| {
            Some(if let Some(first) = first {
                *parent_is_first = first;
                first
            } else {
                *parent_is_first
            })
        })
        .collect()
```